R/fit_model.R

```diff
@@ -0,0 +1,52 @@
+fit_model <- function(data, formula, ...) {
+  UseMethod("fit_model")
+}
+
+#' @export
+fit_model.RtGam_gam <- function(
+    data,
+    formula,
+    family = "nb",
+    method = "REML",
+    ...) {
+  # Override the defaults in formals with the user-supplied args in dots
+  formal_arg_names <- Filter(function(x) x != "...", names(formals()))
+  formals <- as.list(environment())[formal_arg_names]
+  dots <- rlang::list2(...)
+  args <- utils::modifyList(formals, dots)
+
+  do.call(mgcv::gam, args)
+}
+
+#' @export
+fit_model.RtGam_bam <- function(
+    data,
+    formula,
+    family = "nb",
+    method = "fREML",
+    discrete = TRUE,
+    ...) {
+  # Override the defaults in formals with the user-supplied args in dots
+  formal_arg_names <- Filter(function(x) x != "...", names(formals()))
+  formals <- as.list(environment())[formal_arg_names]
+  dots <- rlang::list2(...)
+  args <- utils::modifyList(formals, dots)
+  do.call(mgcv::bam, args)
+}
+
+#' Used to throw informative error if non-supported backend supplied
+#' @export
+fit_model.default <- function(
+    data,
+    formula,
+    ...) {
+  requested_backend <- class(data)[1]
+  supported_backends <- c("gam", "bam")
```

Comment:
can you avoid hard coding this? or if not reword to internal backends?

Reply:
This is a repeat from last time but your response just summaries some tradeoffs without really providing evidence either way.

> I think the counterpoint to consider is what to do if/when we need to move to S3 for other steps (I'm thinking predict and maybe formula would be needed). Centralizing the check for backends into one place adds clarity I'd think, at the cost of extensibility as you noted.

I.e that

I am not sure why this package needs to be non-standard vs the normal approach essentially (which is to let the errors happen per generic)

Reply:
Ha! Fair point.

I think the argument that we shouldn't do non-standard things without a good reason is a compelling one. 

My counterpoint is that I find the default R error message here pretty opaque, especially if you're not super familiar with writing packages. See for example the default error here:

```r
RtGam(
  cases = c(1L, 2L, 3L),
  reference_date = as.Date(c("2023-01-01", "2023-01-02", "2023-01-03")),
  backend = "gan"
)
```

> Error in UseMethod("fit_model") : 
  no applicable method for 'fit_model' applied to an object of class "c('RtGam_gan', 'data.frame')"
Called from: (function(data, formula, ...) {
  UseMethod("fit_model")
})(data = list(cases = 1:3, timestep = c(0, 0.5, 1), reference_date = c(19358, 
19359, 19360), group = c(NA, NA, NA)), formula = cases ~ 1 + 
    s(timestep, k = 3, bs = "tp"))

I think this error is unintuitive. I called `RtGam()` and now it's throwing an error in method dispatch in some internal helper function? The typo `gan` -> `gam` is pretty hard to spot and not clearly the root of the problem.

I would feel differently if the error was more clearly about the backend. But if you think I'm off base here and the extensibility is crucial at this stage, I'll back down. I do think is a bit of an uncomfortable deviation from standard practice.

Reply:
I'm not disagreeing about a error message in default which I like but I don't think the only solution is to hard code the support backends? You can just check what is available using `methods()`.